File: api-gateway-mobile/app/circuit_breaker.py
from __future__ import annotations

import threading
import time
from typing import Any, Callable


class CircuitOpenError(Exception):
    def __init__(self, service: str) -> None:
        self.service = service
        super().__init__(f"Circuit breaker open for {service}")


class CircuitBreaker:
    """CLOSED -> OPEN after consecutive failures; single probe after reset timeout."""

    def __init__(
        self,
        name: str,
        *,
        error_threshold: int = 5,
        reset_timeout_ms: int = 15000,
    ) -> None:
        self._name = name
        self._threshold = max(1, int(error_threshold))
        self._reset_ms = max(1, int(reset_timeout_ms))
        self._lock = threading.Lock()
        self._state = "CLOSED"
        self._failures = 0
        self._opened_at: float | None = None

    def _try_open_to_half_open(self) -> None:
        if self._state != "OPEN" or self._opened_at is None:
            return
        if (time.monotonic() - self._opened_at) * 1000 >= self._reset_ms:
            self._state = "HALF_OPEN"

    def _record_success(self) -> None:
        self._failures = 0
        self._state = "CLOSED"

    def _record_failure(self) -> None:
        if self._state == "HALF_OPEN":
            self._state = "OPEN"
            self._opened_at = time.monotonic()
            self._failures = 0
            return
        self._failures += 1
        if self._failures >= self._threshold:
            self._state = "OPEN"
            self._opened_at = time.monotonic()
            self._failures = 0

    def run(self, fn: Callable[[], Any], *, http_response: bool = False) -> Any:
        with self._lock:
            self._try_open_to_half_open()
            if self._state == "OPEN":
                raise CircuitOpenError(self._name)

        try:
            result = fn()
        except Exception:
            with self._lock:
                self._record_failure()
            raise

        bad_http = False
        if http_response and result is not None:
            status = getattr(result, "status_code", None)
            if isinstance(status, int) and status >= 500:
                bad_http = True

        with self._lock:
            if bad_http:
                self._record_failure()
            else:
                self._record_success()

        return result

File: api-gateway/src/circuitBreaker.js
class CircuitOpenError extends Error {
  constructor(service) {
    super(`Circuit breaker open for ${service}`);
    this.code = "CIRCUIT_OPEN";
    this.service = service;
  }
}

/**
 * Minimal circuit breaker: CLOSED -> OPEN after consecutive failures, half-open probe after reset timeout.
 * @param {{ name: string, errorThreshold?: number, resetTimeoutMs?: number }} opts
 */
function createCircuitBreaker(opts) {
  const name = opts.name;
  const errorThreshold = Math.max(1, Number(opts.errorThreshold) || 5);
  const resetTimeoutMs = Math.max(1, Number(opts.resetTimeoutMs) || 15000);

  let state = "CLOSED";
  let failures = 0;
  let openedAt = null;

  function tryOpenToHalfOpen() {
    if (state !== "OPEN" || openedAt == null) {
      return;
    }
    if (Date.now() - openedAt >= resetTimeoutMs) {
      state = "HALF_OPEN";
    }
  }

  function recordSuccess() {
    failures = 0;
    state = "CLOSED";
  }

  function recordFailure() {
    if (state === "HALF_OPEN") {
      state = "OPEN";
      openedAt = Date.now();
      failures = 0;
      return;
    }
    failures += 1;
    if (failures >= errorThreshold) {
      state = "OPEN";
      openedAt = Date.now();
      failures = 0;
    }
  }

  /**
   * @param {() => Promise<any>} fn
   * @param {{ httpResponse?: boolean }} [runOpts] If httpResponse, status >= 500 counts as failure (axios shape).
   */
  async function run(fn, runOpts = {}) {
    tryOpenToHalfOpen();
    if (state === "OPEN") {
      throw new CircuitOpenError(name);
    }

    try {
      const result = await fn();
      const httpMode = Boolean(runOpts.httpResponse);
      const badHttp =
        httpMode &&
        result &&
        typeof result === "object" &&
        typeof result.status === "number" &&
        result.status >= 500;

      if (badHttp) {
        recordFailure();
        return result;
      }
      recordSuccess();
      return result;
    } catch (err) {
      recordFailure();
      throw err;
    }
  }

  function getState() {
    tryOpenToHalfOpen();
    return { name, state, failures };
  }

  return { run, getState };
}

module.exports = { createCircuitBreaker, CircuitOpenError };
